packages/dashboard/src/DashboardLayout.tsx

```diff
@@ -0,0 +1,254 @@
+  const [lastConfig, setLastConfig] = useState<DashboardLayoutConfig>();
+  const [initialClosedPanels] = useState(data?.closed ?? []);
+
+  const hydrateMap = useMemo(() => new Map(), []);
+  const dehydrateMap = useMemo(() => new Map(), []);
+  const store = useStore();
+  const registerComponent = useCallback(
+    (
+      name: string,
+      componentType: ComponentType,
+      hydrate = hydrateDefault,
+      dehydrate = dehydrateDefault
+    ) => {
+      log.debug2('registerComponent', name, componentType, hydrate, dehydrate);
+
+      function renderComponent(
+        props: { glContainer: unknown; glEventHub: unknown },
+        ref: unknown
+      ) {
+        // Cast it to an `any` type so we can pass the ref in correctly.
+        // ComponentType doesn't seem to work right, ReactNode is also incorrect
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        const CType = componentType as any;
+
+        // Props supplied by GoldenLayout
+        // eslint-disable-next-line react/prop-types
+        const { glContainer, glEventHub } = props;
+        return (
+          <Provider store={store}>
+            <PanelErrorBoundary
+              glContainer={glContainer}
+              glEventHub={glEventHub}
+            >
+              {/* eslint-disable-next-line react/jsx-props-no-spreading */}
+              <CType {...props} ref={ref} />
+            </PanelErrorBoundary>
+          </Provider>
+        );
+      }
+
+      const wrappedComponent = React.forwardRef(renderComponent);
+      const cleanup = layout.registerComponent(name, wrappedComponent);
+      hydrateMap.set(name, hydrate);
+      dehydrateMap.set(name, dehydrate);
+      return cleanup;
+    },
+    [hydrateMap, dehydrateMap, layout, store]
+  );
+  const hydrateComponent = useCallback(
+    (name, props) => hydrateMap.get(name)?.(props, id),
+    [hydrateMap, id]
+  );
+  const dehydrateComponent = useCallback(
+    (name, props) => dehydrateMap.get(name)?.(props, id),
+    [dehydrateMap, id]
+  );
+  const panelManager = useMemo(
+    () =>
+      new PanelManager(
+        layout,
+        hydrateComponent,
+        dehydrateComponent,
+        new Map(),
+        initialClosedPanels,
+        ({ closed, openedMap }) => {
+          dispatch(updateDashboardData(id, { closed, openedMap }));
+        }
+      ),
+    [
+      dehydrateComponent,
+      dispatch,
+      hydrateComponent,
+      id,
+      initialClosedPanels,
+      layout,
+    ]
+  );
+
+  const handleLayoutStateChanged = useCallback(() => {
+    // we don't want to emit stateChanges that happen during item drags or else
+    // we risk the last saved state being one without that panel in the layout entirely
+    if (isItemDragging) return;
+
+    const glConfig = layout.toConfig();
+    const contentConfig = glConfig.content;
+    const dehydratedLayoutConfig = LayoutUtils.dehydrateLayoutConfig(
+      contentConfig
```

Comment:
Should probably pass the `dehydrateComponent` here if we want to be able to use custom dehydrate functions